tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/MkSeqRule.scala

```diff
@@ -0,0 +1,80 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt._
+import at.forsyte.apalache.tla.bmcmt.rules.aux.ProtoSeqOps
+import at.forsyte.apalache.tla.bmcmt.types.IntT
+import at.forsyte.apalache.tla.lir.TypedPredefs._
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.oper.ApalacheOper
+import at.forsyte.apalache.tla.lir.storage.BodyMapFactory
+import at.forsyte.apalache.tla.lir.transformations.impl.IdleTracker
+import at.forsyte.apalache.tla.lir.values.TlaInt
+import at.forsyte.apalache.tla.pp.{ConstSimplifier, InlinerOfUserOper, TlaInputError}
+
+/**
+ * Rewriting rule for MkSeq. This rule is similar to [[FoldSeqRule]].
+ *
+ * @author
+ *   Igor Konnov
+ */
+class MkSeqRule(rewriter: SymbStateRewriter) extends RewritingRule {
+  private val proto = new ProtoSeqOps(rewriter)
+  private val simplifier = new ConstSimplifier(new IdleTracker())
+
+  override def isApplicable(symbState: SymbState): Boolean = {
+    symbState.ex match {
+      // match the internal representation of lambda expressions or embedded calls
+      case OperEx(ApalacheOper.mkSeq, _, LetInEx(NameEx(appName), TlaOperDecl(operName, params, _))) =>
+        appName == operName && params.size == 1
+      case _ => false
+    }
+  }
+
+  override def apply(state: SymbState): SymbState = state.ex match {
+    case OperEx(ApalacheOper.mkSeq, lenEx, LetInEx(NameEx(_), opDecl)) =>
+      val capacity = simplifier(lenEx) match {
+        case ValEx(TlaInt(n)) if n.isValidInt && n >= 0 =>
+          n.toInt
+
+        case unexpectedEx =>
+          val msg = "Expected a constant integer expression. Found: " + unexpectedEx
+          throw new TlaInputError(msg, Some(lenEx.ID))
+      }
+
+      val operT = opDecl.typeTag.asTlaType1()
+      val elemT = operT match {
+        case OperT1(Seq(IntT1()), elemT) => elemT
+        case tp                          =>
+          // this case should be found by the type checker
+          val msg = "Expected an operator of type Int => <elem type>. Found: " + tp
+          throw new TlaInputError(msg, Some(opDecl.ID))
+      }
+
+      // expressions are transient, we don't need tracking
+      val inliner = InlinerOfUserOper(BodyMapFactory.makeFromDecl(opDecl), new IdleTracker)
+
+      def mkElem(state: SymbState, index: Int): (SymbState, ArenaCell) = {
+        // get the cell for the index
+        val (newArena, indexCell) = rewriter.intValueCache.create(state.arena, index)
+        // elem = A(indexCell)
+        val appEx = tla.appOp(tla.name(opDecl.name).as(operT), indexCell.toNameEx.as(IntT1()))
```

Comment:
I was about to suggest using `tla.appDecl`, but then I looked into the source of that method, and I think it wouldn't correctly propagate the type tag in line https://github.com/informalsystems/apalache/blob/51f073a6029de473db2c19676f628f2120ed55f2/tlair/src/main/scala/at/forsyte/apalache/tla/lir/Builder.scala#L186

Can you check out `tla.appDecl` and see if it correctly types the `NameEx`? If not, I think we need to open an issue for all compound methods in the builder (`appDecl`, `primeEq`, `unchangedTup`, etc.) (unrelated to this PR overall).

Reply:
yes, I don't think `appDecl` ever worked correctly with types.